### ISSUE TITLE
Repeated errors of: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout

### ISSUE THREAD

**Issue by calebpitts** [polygon-io/client-python#472]
**Describe the bug**
After subscribing to all equity trades, I keep getting my websocket connection disconnected with this error:
`connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received`

**To Reproduce**
```
stream_task = asyncio.create_task(self.polygon_websocket_client.connect(self.message_handler.add))
handle_task = asyncio.create_task(self.message_handler.start_handling())
asyncio.gather(stream_task, handle_task)
```

```
async def add(self, message_response: Optional[Union[str, bytes]]) -> None:
        await self.handler_queue.put(message_response)  # producer
```

```
async def start_handling(self) -> None:
        while True:
            message_response = await self.handler_queue.get()
            # handle message here
```

Running this, will produce the 1011 error after anywhere between 40 seconds to 5 minutes. 

**Expected behavior**
I'm immediately putting all messages into a processing queue to avoid server-side buffer latency disconnections. I'd expect that my connection is not disconnected at all throughout the trading time until my script disconnects at the end of the trading day (which is tracked by a separate async process I have running)

Anyone else have this issue?

**Comment by justinpolygon** [comment]
Hey @calebpitts, are you seeing any trades come through at all? Basically, I'm wanting to know if it works for a bit and then stops? Do you mind posting the compete code sample so that I can run it.

We have an example here mind trying to run this https://github.com/polygon-io/client-python/blob/master/examples/websocket/stocks-ws.py with `client.subscribe("T.*")` uncommented.

There should also be something on the dashboard that tells you about the connection https://polygon.io/dashboard/connections mind checking that out?

**Comment by calebpitts** [comment]
Hi @justinpolygon,

I'm able to stream trades, and the client does a good job at reconnecting, but the problem is that it disconnects every few minutes, making me miss data and if I don't set a high max_reconnect value, the client will stop trying. 

My lifecycle of events looks exactly like you pictured (time between subscribed and client disconnected is 40secs-5mins). 

Here's a stripped down version of what I am doing (apologies if there are inconsistent spacing):
```
import asyncio
from polygon import RESTClient, WebSocketClientfrom 
polygon.websocket.models import Market


class MessageHandler:
    def __init__(rest_client):
         self.rest_client = rest_client
         self.handler_queue = asyncio.Queue()
   
    async def add(self, message_response: Optional[Union[str, bytes]]) -> None:
        await self.handler_queue.put(message_response)  # producer

     async def start_handling(self) -> None:
        while True:
            message_response = await self.handler_queue.get()
            # handle message here
            print("HANDLING:", message_response)
        

class MyClient:
    def __init__(feed, market):
        self.polygon_websocket_client = polygon_websocket_client or WebSocketClient(
            api_key="XXXXXXX",
            feed=feed,
            market=market,
            raw=True,
            verbose=True,
            subscriptions=["T.*"],
        )
      self.polygon_rest_client = polygon_rest_client or RESTClient(api_key="XXXXXX")
      self.message_handler = MessageHandler(self.polygon_rest_client)

    async def start_event_stream(self):
        await asyncio.create_task(self.polygon_websocket_client.connect(self.message_handler.add))

async def main():
    my_client = MyClient(
            feed=Feed.RealTime,
            market=Market.Options,
    )
    asyncio.gather(my_client.start_event_stream())

asyncio.run(main())
```

**Comment by calebpitts** [comment]
after some digging, perhaps I need to mess with the `ping_timeout` setting and set it to None? https://websockets.readthedocs.io/en/stable/topics/timeouts.html

I'll try this tomorrow.

**Comment by justinpolygon** [comment]
Hey @calebpitts, I've looked at the code you shared, and I wanted to share a couple things however I don't think if that'll solve your problem but it's worth testing. Looks like the script connects to both the rest client and the websocket. For debugging I'd probably suggest just trying to only connect to the websocket until we nail down the issue. I do the same thing, connect to both, but just wanted to make sure we scope it down so we know what calls what and debug the issue.

I think Python's asyncio tasks need to be awaited to ensure they run as expected. In the current script, there are some asyncio tasks that aren't explicitly awaited, which might cause unexpected behavior or even result in some tasks not running at all. For instance, in the start_event_stream method, the created task should ideally be awaited.

Here's a corrected version of that part:

```python
async def start_event_stream(self):
    await self.polygon_websocket_client.connect(self.message_handler.add)
```

Additionally, in the main function, the asyncio.gather() function should also be awaited:

```python
async def main():
    my_client = MyClient(
            feed=Feed.RealTime,
            market=Market.Options,
    )
    await asyncio.gather(my_client.start_event_stream())
```

Here's the updated script with some other tweaks:

```python
import asyncio
from typing import Optional, Union
from polygon import WebSocketClient
from polygon.websocket.models import Market, Feed

class MessageHandler:
    def __init__(self):
        self.handler_queue = asyncio.Queue()

    async def add(self, message_response: Optional[Union[str, bytes]]) -> None:
        await self.handler_queue.put(message_response)  # producer

    async def start_handling(self) -> None:
        while True:
            message_response = await self.handler_queue.get()
            print("HANDLING:", message_response)


class MyClient:
    def __init__(self, feed, market, polygon_websocket_client=None):
        self.polygon_websocket_client = polygon_websocket_client or WebSocketClient(
            api_key="XXXXXX",
            feed=feed,
            market=market,
            raw=True,
            verbose=True,
            subscriptions=["T.*"],
        )
        self.message_handler = MessageHandler()

    async def start_event_stream(self):
        await asyncio.gather(
            self.polygon_websocket_client.connect(self.message_handler.add),
            self.message_handler.start_handling()
        )


async def main():
    my_client = MyClient(
            feed=Feed.RealTime,
            market=Market.Options,
    )
    await asyncio.gather(my_client.start_event_stream())

asyncio.run(main())
```

It might be worth a shot to test with these changes and see if it's more reliably by ensuring all tasks are scheduled and executed. Streaming all trades does send over a significant amount of data, especially at market open and close, and you'll need to have the bandwidth, processing power, etc to make it work. So, it might just be that it's too much data but I'd like to rule a few things out first.

Secondly, it appears the current script is set up to subscribe to options data (as defined by Market.Options). If your intention was to subscribe to stock data instead, you might want to replace Market.Options with Market.Stocks?

I mostly just maintain the library here but we might need to pull in support if this doesn't fix things up. For the connection on the https://polygon.io/dashboard/connections are you seeing messages buffered at all? If this is still happening tomorrow I'll pass things over to the support team.

**Comment by justinpolygon** [comment]
Hey @calebpitts, I've tested the script above listening to all `T.*` trades for the `Market.Options` feed and it's worked for a few hours now. Let me know how things work out. Attached is a screenshot.

![CleanShot 2023-07-19 at 09 25 26](https://github.com/polygon-io/client-python/assets/123573436/b6136503-37f7-4b30-b4a2-5e11ccf1c57e)

**Comment by calebpitts** [comment]
It's doing better, but still got a 1011 error after the volume of options trades picked up:
<img width="951" alt="image" src="https://github.com/polygon-io/client-python/assets/19628946/6ba1d7cb-0600-49b0-a76f-e81606562c27">

I tried with your modification without my extra logic and I'm not seeing any issues. But, if I reintroduce my logic with REST api calls, I'll see the above issue during high volume.

I'm creating a task whenever I process a message off my queue like this:

```
async def start_handling(self) -> None:
       handler_tasks: List[asyncio.Task] = []
        while True:
            message_response = await self.handler_queue.get()
            asyncio.create_task(self._handle_response(message_response))
        await asyncio.gather(*handler_tasks)

async def _handle_response(self, message_response: Union[str, bytes]) -> None:
      # make REST calls to get option quote, open interest, ect.
     pass
```

I thought that by offloading the messages onto a local queue, and running separate tasks on each of those messages as they enter the queue, would allow me to make the REST calls I want to make when processing trades. I'm not making REST calls on all trades, only large ones, so about 3 per second.

**Comment by calebpitts** [comment]
Seeing lots more toward the end of the day:
<img width="882" alt="image" src="https://github.com/polygon-io/client-python/assets/19628946/d053a09e-2198-4d36-9df8-6e1f939897e8">

How would you recommend I handle messages? I'd like to annotate each trade with the quote at the time of the trade and I was doing that in a separate task after popping off the message queue. That way I could run all my quote requests concurrently. I don't care about each task taking about a second. I don't want to do it synchronously though, as that would add a ton of runtime time toward the end of the day waiting for each REST response before send another REST request. 

Result of a grep command looking for the error:
```
2023-07-19 12:30:10,718 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
2023-07-19 12:30:54,257 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
2023-07-19 12:31:37,243 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
2023-07-19 12:32:19,792 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
2023-07-19 12:41:34,813 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
2023-07-19 12:42:17,611 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
2023-07-19 12:43:07,820 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
2023-07-19 12:45:21,568 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
2023-07-19 12:46:30,755 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
2023-07-19 12:48:37,032 WebSocketClient DEBUG: connection closed (ERR): sent 1011 (unexpected error) keepalive ping timeout; no close frame received
```

**Comment by justinpolygon** [comment]
Hey @calebpitts. This is great, at least we've narrowed it down to the REST calls and that you're able to handle the stream of data without doing them. So, that's something at least. I wonder if at peak times, when there are lots of messages coming in, your app is making more calls. If each of these takes a second or more to complete, they might stack up and be the cause of the issues here. What are you doing after you get that response from the REST API?

My suggestion here is to make sure to handle the tasks created within the `start_handling` method as quickly as possible, ideally not blocking at all. You might even want to profile these tasks with a start and an end time, then diff them to see a total time taken for each task, just so you know for sure what's going on in there. While creating a task with `asyncio.create_task()` schedules it to run, there might be potential issues if these tasks are not being awaited or managed correctly. They could be cleaned up prematurely, or if an exception is raised within one of these tasks, it could go unnoticed, or cause your app to stop.

Here's an adjusted version of the `start_handling` method that appends the created tasks to a list, and cleans up finished tasks at each iteration to avoid potential memory issues:

```python
async def start_handling(self) -> None:
    handler_tasks: List[asyncio.Task] = []
    while True:
        message_response = await self.handler_queue.get()
        task = asyncio.create_task(self._handle_response(message_response))
        handler_tasks.append(task)

        # Clean up finished tasks
        handler_tasks = [t for t in handler_tasks if not t.done()]
    await asyncio.gather(*handler_tasks)
```

This suggestion is much more complex, but let's take it to the extreme for a minute. Say you needed to handle 100k+ messages a second. How would you do that? Depending on the load you are handling after making the API calls, you might want to consider a setup involving separate programs for the WebSocket listener (producer) and tasks making the API calls (consumers), possibly with a message queueing system like RabbitMQ in between. Then you can scale up the consumers as needed to handle the load.

WebSocket listener (producer) -> RabbitMQ -> tasks making the API calls (consumers)

Basically, you would have one python script listening to the WebSocket, a RabbitMQ server acting as the queue (or Kafka was beyond what you'd need here), and then several python consumer scripts reading messages off the queue that aren't in the same script as the WebSocket. This could help if a single Python script can't handle the volume of data and processing you're dealing with.

This approach will definitely work for you if you cannot scale within this single program. However, it is much more complex and involves much more overhead than your current setup. I'd probably try the updated start_handling, profile your jobs, and see exactly how long they are taking.

**Comment by calebpitts** [comment]
thanks @justinpolygon I started clearing tasks and switched over to subscribing on all stock trades instead since the market closed and options were no longer trading. Got better results:
<img width="1019" alt="image" src="https://github.com/polygon-io/client-python/assets/19628946/2c5f4e39-79c2-41fe-aa90-53ce20bc5dd6">

I'm doing more REST calls on option trades however, so I'll test this approach out tomorrow. Also working on handling any errors that occur, there's a good chance an error might be blocking the acceptance of new trades. 

And thanks for the RabitMQ suggestion. Hopefully handling errors and cleaning up finished tasks fixes this, I'll report back tomorrow after getting a run in.

**Comment by justinpolygon** [comment]
Great, sounds good. Thanks for letting me know. Good luck.

**Comment by calebpitts** [comment]
Unfortunately still getting the same error with clearing pending tasks and encapsilating my handler in a try/except. With your last example inside `_handle_response()` are you able to make 2-3 REST calls for trades with a total cost / premium over 10k? I'm collecting the quote and spot price. But you might be able to recreate if you just make two duplicate quote REST calls. Still looking through the disconnects now to see if there is a pattern, but haven't spotted anything yet. 

<img width="528" alt="image" src="https://github.com/polygon-io/client-python/assets/19628946/bcb5bed0-4480-4f6b-91a8-99ce57288d3b">

Going to try one more time tomorrow. This likely is an issue on my end, but want to make sure there is no issue with using the REST and WebSocket clients in the same async tasks.

**Comment by calebpitts** [comment]
@justinpolygon I think I have good news. I was able to reproduce the error in my own testing with a fake server and spammed 30k messages to my client. I managed to isolate the error down to the two API calls I was making within `_handle_response`. If I comment those two out, the error went away. I did some digging and I found that if I used `asyncio.to_thread` and nested the two rest call functions in a synchronous function, the error went away. 

Example:
```
async def _handle_response(self, message_response: Union[str, bytes]) -> None:
    message_json: List[Dict[str, Any]] = json.loads(message_response)
    async for message in MessageParser.parse_all(message_json):
        await asyncio.to_thread(self._handle_message, message)

def _handle_message(self, message) -> None:
      # execute rest call 1 and 2 and then build an object that I upload to my db
```

`parse_all` is my own parser function that parses and immediately yields each message in the message group provided by the websocket asynchronously. 

Going to try this with the real data tomorrow.

**Comment by calebpitts** [comment]
I believe that was the issue. Drop counts went down, going to refactor my code some more. Thanks for the help.

<img width="494" alt="image" src="https://github.com/polygon-io/client-python/assets/19628946/359e3c14-59dc-4768-a2c1-88e3cabe0524">

**Comment by justinpolygon** [comment]
Amazing, I'm so happy to hear it worked out! Cheers @calebpitts. Have a great weekend!
